File: test_commit_files.py
import os
import subprocess
import tempfile
import shutil
import unittest
import json

class TestCommitFiles(unittest.TestCase):
    def setUp(self):
        # Create a temporary directory
        self.test_dir = tempfile.mkdtemp()
        
        # Initialize a git repository
        subprocess.run(['git', 'init', self.test_dir], check=True)
        
        # Change to the test directory
        os.chdir(self.test_dir)
        
        # Set up git config
        subprocess.run(['git', 'config', 'user.email', 'test@example.com'], check=True)
        subprocess.run(['git', 'config', 'user.name', 'Test User'], check=True)

    def tearDown(self):
        # Change back to the original directory
        os.chdir(os.path.dirname(os.path.realpath(__file__)))
        
        # Remove the temporary directory
        shutil.rmtree(self.test_dir)

    def create_test_file(self, filename, content):
        with open(os.path.join(self.test_dir, filename), 'w') as f:
            f.write(content)

    def run_commit_script(self, script_name):
        if script_name.endswith('.py'):
            result = subprocess.run(['python3', script_name], capture_output=True, text=True)
        elif script_name.endswith('.php'):
            result = subprocess.run(['php', script_name], capture_output=True, text=True)
        else:
            raise ValueError(f"Unsupported script type: {script_name}")
        return result.stdout

    def test_commit_files(self):
        # Create test files
        self.create_test_file('test1.txt', 'Author: John Doe\nThis is test file 1\n#tag1 #tag2')
        self.create_test_file('test2.txt', 'Author: Jane Smith\nThis is test file 2\n#tag3')
        self.create_test_file('not_a_text_file.dat', 'This should be ignored')

        # Copy the commit scripts to the test directory
        shutil.copy('commit_files.py', self.test_dir)
        shutil.copy('commit_files.php', self.test_dir)

        # Run both scripts
        py_output = self.run_commit_script('commit_files.py')
        php_output = self.run_commit_script('commit_files.php')

        # Check if both scripts produced similar output
        self.assertIn("Committed 2 text files", py_output)
        self.assertIn("Committed 2 text files", php_output)

        # Check if metadata files were created
        metadata_dir = os.path.join(self.test_dir, 'metadata')
        self.assertTrue(os.path.exists(metadata_dir))
        self.assertEqual(len(os.listdir(metadata_dir)), 2)

        # Check metadata content
        for filename in ['test1.txt.json', 'test2.txt.json']:
            with open(os.path.join(metadata_dir, filename), 'r') as f:
                metadata = json.load(f)
                self.assertIn('author', metadata)
                self.assertIn('title', metadata)
                self.assertIn('hashtags', metadata)
                self.assertIn('file_hash', metadata)

        # Check git log
        result = subprocess.run(['git', 'log', '--oneline'], capture_output=True, text=True)
        self.assertIn("Auto-commit 2 text files", result.stdout)

if __name__ == '__main__':
    unittest.main()

File: test_commit_files.js
// test_commit_files.js
// to run: node test_commit_files.js

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

function randomString(length = 8) {
    return crypto.randomBytes(length).toString('hex');
}

function createTestFile(content) {
    const dateStr = new Date().toISOString().split('T')[0];
    const filename = `${randomString()}.txt`;
    const dirPath = path.join('message', dateStr);
    fs.mkdirSync(dirPath, { recursive: true });
    const filePath = path.join(dirPath, filename);
    
    fs.writeFileSync(filePath, content);
    
    return filePath;
}

function runCommitFiles(script) {
    execSync(script, { stdio: 'inherit' });
}

function checkGitLog() {
    const logOutput = execSync('git log -1 --pretty=format:%s', { encoding: 'utf-8' });
    return logOutput.includes('Auto-commit');
}

function checkMetadataFile(filename) {
    const metadataFile = path.join(path.dirname(filename), 'metadata', path.basename(filename) + '.json');
    if (!fs.existsSync(metadataFile)) {
        return false;
    }
    const metadata = JSON.parse(fs.readFileSync(metadataFile, 'utf-8'));
    return ['author', 'title', 'hashtags', 'file_hash'].every(key => key in metadata);
}

function runTests(script) {
    console.log(`Testing ${script}`);

    // Test 1: Commit a single file
    const file1 = createTestFile("Author: John Doe\nThe Beauty of Nature\n\nNature's beauty is an awe-inspiring spectacle that never fails to amaze us. From the grandeur of mountains to the delicacy of a flower, it reminds us of the world's magnificence.\n\n#nature #beauty #inspiration");
    runCommitFiles(script);
    if (!checkGitLog()) throw new Error('Git commit not found');
    if (!checkMetadataFile(file1)) throw new Error('Metadata file not created or invalid');
    console.log('Test 1 passed: Single file commit');

    // Test 2: Commit multiple files
    const file2 = createTestFile("Author: Jane Smith\nThe Art of Cooking\n\nCooking is not just about sustenance; it's an art form that engages all our senses. The sizzle of a pan, the aroma of spices, and the vibrant colors of fresh ingredients all come together to create culinary masterpieces.\n\n#cooking #art #food");
    const file3 = createTestFile("Author: Bob Johnson\nThe Joy of Learning\n\nLearning is a lifelong journey that opens doors to new worlds. Whether it's picking up a new skill or diving deep into a subject, the process of discovery and growth is incredibly rewarding.\n\n#learning #education #growth");
    runCommitFiles(script);
    if (!checkGitLog()) throw new Error('Git commit not found');
    if (!checkMetadataFile(file2) || !checkMetadataFile(file3)) throw new Error('Metadata files not created or invalid');
    console.log('Test 2 passed: Multiple file commit');

    // Test 3: No changes to commit
    runCommitFiles(script);
    console.log('Test 3 passed: No changes to commit');

    console.log(`All tests passed for ${script}`);
}

const scripts = [
    'python3 commit_files.py',
    'node commit_files.js',
    'php commit_files.php',
    'perl commit_files.pl',
    'ruby commit_files.rb'
];

scripts.forEach(runTests);
